bc_obps/service/facility_service.py

```diff
@@ -50,36 +166,61 @@ def get_if_authorized(cls, user_guid: UUID, facility_id: UUID) -> Facility:
-            raise Exception(UNAUTHORIZED_MESSAGE)
-        # if any of address data generate one address and keep it to pass to the facility
-        facility_data: dict = payload.dict(
-            include={
-                'name',
-                'type',
-                'latitude_of_largest_emissions',
-                'longitude_of_largest_emissions',
-            }
-        )
-        address_data = payload.dict(
-            include={'street_address', 'municipality', 'province', 'postal_code'}, exclude_none=True
-        )
-        if address_data:
-            address = AddressDataAccessService.create_address(address_data)
+        cls.check_user_access(user_guid, operation)
+
+        facility_data = cls.prepare_facility_data(payload)
+        address = cls.create_address(payload)
+        if address:
             facility_data['address'] = address
+
         facility = FacilityDataAccessService.create_facility(user_guid, facility_data)
         FacilityOwnershipTimelineDataAccessService.create_facility_ownership_timeline(
             user_guid, {'facility': facility, 'operation': operation, 'start_date': timezone.now()}
         )
 
-        well_authorization_numbers = []
-        for number in payload.well_authorization_numbers:
-            well_authorization_numbers.append(
-                WellAuthorizationNumberDataAccessService.create_well_authorization_number(user_guid, number)
-            )
-        if well_authorization_numbers:
-            facility.well_authorization_numbers.set(well_authorization_numbers)
+        cls.handle_well_authorization_numbers(user_guid, payload, facility)
+
+        return facility
+
+    @classmethod
+    @transaction.atomic()
+    def update_facility(cls, user_guid: UUID, facility_id: UUID, payload: FacilityIn) -> Facility:
+        """
+        Checks user access and, if authorized, updates a facility with new data.
+
+        Parameters:
+        - user_guid (UUID): The GUID of the user making the update request.
+        - facility_id (UUID): The ID of the facility to be updated.
+        - payload (FacilityIn): The new data for updating the facility.
+
+        Returns:
+        - Facility: The updated facility instance.
+        """
+
+        # Retrieve the facility object using the provided facility ID
+        facility: Facility = FacilityDataAccessService.get_by_id(facility_id)
+
+        # Retrieve the operation object using the operation ID from the payload
+        operation = OperationDataAccessService.get_by_id(payload.operation_id)
+
+        # Check if the user has access to update the given operation
+        cls.check_user_access(user_guid, operation)
+
+        # Prepare the facility data for updating, based on the provided payload
+        facility_data = cls.prepare_facility_data(payload)
+
+        # Create or update the address associated with the facility, if provided
+        address = cls.create_address(payload, False)
```

Comment:
I ~think this will create a new address record every time. We need to create one if it wasn't previously entered, but otherwise it's an update